checker/src/org/checkerframework/checker/index/upperbound/UpperBoundTransfer.java

```diff
@@ -124,14 +131,30 @@ private void knownToBeArrayLength(
 
         UBQualifier newInfo = UBQualifier.createUBQualifier(arrayExp, "-1");
         UBQualifier newLeft = left.glb(newInfo.plusOffset(node.getRightOperand(), atypeFactory));
+        newLeft = accountForLowerBoundAnnos(node.getRightOperand(), newLeft, arrayExp);
+
         Receiver leftRec = FlowExpressions.internalReprOf(atypeFactory, node.getLeftOperand());
         store.insertValue(leftRec, atypeFactory.convertUBQualifierToAnnotation(newLeft));
 
         UBQualifier newRight = right.glb(newInfo.plusOffset(node.getLeftOperand(), atypeFactory));
+        newRight = accountForLowerBoundAnnos(node.getLeftOperand(), newRight, arrayExp);
+
         Receiver rightRec = FlowExpressions.internalReprOf(atypeFactory, node.getRightOperand());
         store.insertValue(rightRec, atypeFactory.convertUBQualifierToAnnotation(newRight));
     }
 
+    /** If the node is NN but not POS, add an LTEL to the qual. If POS, an LTL. */
+    private UBQualifier accountForLowerBoundAnnos(Node node, UBQualifier qual, String arrayExp) {
+        if (atypeFactory.hasLowerBoundTypeByClass(node, NonNegative.class)) {
+            if (atypeFactory.hasLowerBoundTypeByClass(node, Positive.class)) {
+                qual = qual.glb(UBQualifier.createUBQualifier(arrayExp, "0"));
```

Comment:
I'll refactor this to be better. I agree with you that it's wrong. These should just be separate cases at the top level.